### PR TITLE
Implement Hype forcing & model

### PR DESCRIPTION
**Pull request by sverhoeven**
TODO
* [x] HypeForcing
* [x] Hype model
* [x] In docs
* [x] Check docs/adding_models.rst steps have been done
* [ ] ~Add https://sourceforge.net/projects/hype/files/release_hype_5_6_2/demo.zip as example parameter set~
* [x] use model.time_as_isostr, but model.bmi.get_time_units() returns invalid 'hours since start of simulation'
* [ ] ~find demo set where spatial position is known of sub-basins and then use model.get_value_as_xarray()~
* [x] Fix for Python 3.7 -> lets drop 3.7 support
* [ ] Copy sif to dcache

### CODE REVIEW

**Comment by SarahAlidoost** [comment]
@sverhoeven the notebook [hype.ipynb](https://github.com/eWaterCycle/ewatercycle/blob/0594766b783b21f7cbad05ec77731c82d7c95239/docs/examples/hype.ipynb) using docker gives the error:

```python
---------------------------------------------------------------------------
DeadContainerException                    Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 cfg_file, cfg_dir = model.setup()
      2 cfg_file, cfg_dir

File ~/GitHub/ewatercycle/src/ewatercycle/models/hype.py:152, in Hype.setup(self, start_time, end_time, crit_time, cfg_dir)
    150 # start container
    151 work_dir = str(cfg_dir_as_path)
--> 152 self.bmi = _start_container(self.version, work_dir)
    154 return str(cfg_file), work_dir

File ~/GitHub/ewatercycle/src/ewatercycle/models/hype.py:217, in _start_container(version, work_dir)
    215 elif CFG["container_engine"].lower() == "docker":
    216     image = _version_images[version]["docker"]
--> 217     return BmiClientDocker(
    218         image=image,
    219         image_port=55555,  # TODO needed?
    220         work_dir=work_dir,
    221     )
    222 else:
    223     raise ValueError(
    224         f"Unknown container technology in CFG: {CFG['container_engine']}"
    225     )

File ~/mambaforge/envs/ewatercycle/lib/python3.10/site-packages/grpc4bmi/bmi_client_docker.py:88, in BmiClientDocker.__init__(self, image, work_dir, image_port, host, input_dirs, user, remove, delay, timeout)
     86         logs = self.logs()
     87         msg = f'Failed to start Docker container with image {image}, Container log: {logs}'
---> 88         raise DeadContainerException(msg, exitcode, logs)
     90 super(BmiClientDocker, self).__init__(BmiClient.create_grpc_channel(port=port, host=host), timeout=timeout)

DeadContainerException: Failed to start Docker container with image ewatercycle/hype-grpc4bmi:feb2021, Container log: terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
```

I submitted issue eWaterCycle/ewatercycle-hype#5

File: src/ewatercycle/forcing/_hype.py [hunk 1]
```diff
@@ -64,19 +75,22 @@ def generate(  # type: ignore
 
         # generate forcing data and retreive useful information
         recipe_output = recipe.run(session=_session(directory))
-        # TODO return files created by ESMValTOOL which are needed by Hype Model
-        # forcing_path = list(recipe_output['...........']).data_files[0]
-        forcing_path = "/foobar.txt"
 
-        forcing_file = Path(forcing_path).name
```
Comment (SarahAlidoost) [inline src/ewatercycle/forcing/_hype.py:71]: the `from pathlib import Path` at the top of the module can be removed. Shall we also add **Hype generate forcing** (e.g. the code block below) to the notebook [generate_forcing.ipynb](https://github.com/eWaterCycle/ewatercycle/blob/main/docs/examples/generate_forcing.ipynb)? 

```python
# Hype
hype_forcing = ewatercycle.forcing.generate(
    target_model="hype",
    dataset="ERA5",
    start_time="1990-01-01T00:00:00Z",
    end_time="1990-12-31T00:00:00Z",
    shape="./data/Rhine/Rhine.shp",
)
print(hype_forcing)
```

File: docs/examples/hype.ipynb [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "parameter_set_dir = \"<path where demo.zip was extracted to>\"\n",
+    "parameter_set_dir = \"/tmp/o/demo\"\n",
```
Comment (SarahAlidoost) [inline docs/examples/hype.ipynb]:
```suggestion
```

File: tests/models/test_hype.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    )
+
+
+class TestWithOnlyParameterSetAndDefaults:
+    @pytest.fixture
+    def model(self, parameter_set):
+        return Hype("feb2021", parameter_set)
+
+    @pytest.fixture
+    def model_with_setup(self, mocked_config, model: Hype):
```
Comment (SarahAlidoost) [inline tests/models/test_hype.py:81]: unused `mocked_config`?!

**Review by SarahAlidoost — APPROVED** [review]
@sverhoeven I tested the notebook on Snellius. It works. I left some minor comments. We need to add "Hype" to [example.rst](https://github.com/eWaterCycle/ewatercycle/blob/main/docs/examples.rst) as well. Similar to other notebooks, we may run the model for a short time period and plot the time series in the notebook hype.ipynb.

File: src/ewatercycle/forcing/_hype.py (hunk 1 above)
Comment (sverhoeven) [reply src/ewatercycle/forcing/_hype.py:71]: Added, including the print output

**Comment by sverhoeven** [comment]
> @sverhoeven I tested the notebook on Snellius. It works. I left some minor comments. We need to add "Hype" to [example.rst](https://github.com/eWaterCycle/ewatercycle/blob/main/docs/examples.rst) as well. Similar to other notebooks, we may run the model for a short time period and plot the time series in the notebook hype.ipynb.

Added hype to examples.rst and discharge is plotted now.

**Comment by SarahAlidoost** [comment]
esmvaltool issue https://github.com/ESMValGroup/ESMValTool/issues/2678

**Comment by sverhoeven** [comment]
thanks for reviewing. I spawned some additional issues eWaterCycle/ewatercycle-hype#6, eWaterCycle/ewatercycle-hype#4

**Pull request merged**